### ISSUE TITLE
'Host Check Error' crash on startup since upgrading to Syncthing v0.14.6

### ISSUE THREAD

**Issue by ghalfacree** [syncthing/syncthing-gtk#321]
Upgraded from Syncthing v0.14.5 to v0.14.6 this morning, and now Syncthing-GTK crashes on start-up with a 'Host Check Error' message.

```
Server response:
'HTTP/1.0 403 Forbidden
Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Tue, 06 Sep 2016 08:05:06 GMT
Content-Length: 17

Host check error
'
```

My Syncthing instance is password protected, which accounts for the 403 error - but it's always been password protected, and that's never caused Syncthing-GTK any problems in the past. If I start Syncthing-GTK and allow it to bring up the daemon, it appears to do so then pops up the error dialogue and shuts everything down when the dialogue is closed; if I start Syncthing manually then start Syncthing-GTK, the error dialogue still appears but Syncthing continues to run normally when Syncthing-GTK closes.

Debug output follows, though I can see nothing particularly useful. The referenced config.xml contains the correct username and hashed password, and I can access the web UI fine using those credentials.

```
D UIBuilder     Enabled: traditional_header
D UIBuilder     Enabled: old_gtk
D UIBuilder     Enabled: icons_in_menu
D UIBuilder     Loading glade file /usr/share/syncthing-gtk/app.glade
D UIBuilder     Enabled conditions: set([u'old_gtk', u'icons_in_menu', u'traditional_header'])
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Allowed node !is_gnome | old_gtk
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'property' by attribute: !old_gtk
D UIBuilder     Removed 'child' by attribute: header_bar
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'child' by attribute: is_windows
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Allowed node traditional_header
D UIBuilder     Allowed node traditional_header | icons_in_menu
W StatusIcon    StatusIcon: Using backend StatusIconAppIndicator (fallback)
D Daemon        Reasing syncthing config /home/me/.config/syncthing/config.xml
V Notifications Error notifications enabled
V Notifications File update notifications enabled
V Notifications Folder notifications enabled
I App           
I App           Syncthing-GTK started and running in notification area
V Daemon        Set refresh interval to 5
```

**Comment by kozec** [comment]
@calmh sorry for probably stupid question, but what exactly [is this](https://github.com/syncthing/syncthing/commit/49910a1d853a8c392e5faa00c3c2aee0fc060d78#diff-08cb9d814d52dd7eb886f63a0fe1b847R316) supposed to do?

Because what it actually does is breaking ST-GTK right now :(

**Comment by kozec** [comment]
By the way, in theory, setting GUI listen address to 0.0.0.0 should workaround about this. Except that's really horrible idea :)

**Comment by ghalfacree** [comment]
@kozec Yeah, I think I'll probably wait for a proper fix rather than open the web GUI to all and sundry! Nice work tracking down the change in Syncthing that's causing the problem so quickly - fingers crossed it's not too hairy to sort out.

**Comment by calmh** [comment]
@kozec If the interface is bound to 127.0.0.1, the `Host` header should be something like 127.0.0.1, localhost, etc. I don't see how it could possibly be a 10-address?

**Comment by calmh** [comment]
But in the meantime, there is an advanced config, set `insecureSkipHostcheck` to true in the GUI section.

**Comment by ghalfacree** [comment]
I can confirm that setting `insecureSkipHostcheck` in Syncthing's config works to bring Syncthing-GTK back - thanks @calmh.

**Comment by kozec** [comment]
@calmh In ST-GTK case, interface and local address is chosen by glib (or by something even lower) and I'm not 100% sure if it is possible to override it. Why shouldn't be 10.* address considered local?

**Comment by calmh** [comment]
I'm not sure what you mean by chosen by glib. If the listen address is 127.0.0.1:8384, you should access it as http://127.0.0.1:8384/, which should set a `Host: 127.0.0.1:8384` header, which passes the check. How *do* you communicate with Syncthing?

**Comment by calmh** [comment]
I'll download it and try :)

**Comment by calmh** [comment]
Ah. It looks like you're building HTTP 1.0 requests yourself in strings...? That's... not going to work with the default setup. :)

**Comment by calmh** [comment]
Or, well, it'll work if you add a Host header but it seems a fragile way to say the least to speak HTTP. I mean there are excellent HTTP libraries for Python, and HTTP/1.1 has been out for twenty years or so...

**Comment by kozec** [comment]
> I mean there are excellent HTTP libraries for Python, and HTTP/1.1 has been out for twenty years or so...

Well, using any other library is out of question, because that would involve either thread or having UI locked. And thread is sure way to freeze GTK on Windows :(

Also, although glib has HTTP library, it doesn't provide POST method and causes horrible memory leaks on guess-which-os :D

Anyway, are you saying that just pretending that I'm on localhost would work? Because, yeah, that's easy fix and I'll test it right after I get home, but it also makes that check in Syncthing even more pointless. If ST-GTK can send it, attacker can as well.

**Comment by calmh** [comment]
You don't need to pretend anything, you just need to set a correct Host header. You already do that in the request that fetches the CSRF token, I think you just need to add it on the other path as well.

**Comment by calmh** [comment]
An attacker would typically come in via a DNS rebinding attack on a browser, where they do not have power of the host header. If the attacked can create raw sockets they are in another position, but that depends on already having compromised something else.

**Comment by kozec** [comment]
Good news everyone, it works :)

I'll add new release with this fix.

**Comment by bungabunga** [comment]
@kozec is a workaround by enabling insecureSkipHostcheck until this version comes to ubuntu webupd8 repo ok security-vise?

**Comment by kozec** [comment]
@bungabunga Not really, but it shouldn't be more insecure than using Syncthing v0.14.5 was.

**Comment by ghalfacree** [comment]
Replying to a closed issue (naughty-naughty), but just wanted to confirm that the new release dropped onto my box this morning and now Syncthing-GTK works fine with Syncthing v0.14.6 without the need for `insecureSkipHostcheck`. Cheers, @kozec!

**Comment by fraternl** [comment]
I upgraded to 0.14.7 and I get this error message.
I'm connecting through a proxy.

I added:

<insecureSkipHostcheck>true</insecureSkipHostcheck>
/home/syncthing/.config/syncthing/config.xml between <options> and </options>

But that didn't work

**Comment by fraternl** [comment]
I got it to work by enabling the option on another install using the GUI
It needed to be in the gui-section and it is case-sensitive.

Only after doing a copy paste of that section in config.xml I had it going again

        {insecureSkipHostcheck}true{/insecureSkipHostcheck}

**Comment by MrtimH** [comment]
Y

**Comment by gsantner** [comment]
I tried both, adding insecureSkipHostcheck, and editing host in apache reverse proxy config. 
**Neither of it worked.** Still `Host check error` in Web-Gui.

Debian 8 with Syncthing Apt repository, latest available version 0.14.10 installed.
If something on apache config needs to be modified, than the docu should also get updated..

**Comment by spacemanspiff2007** [comment]
I installed via chocolatey.
For completeness:
Creating a new node with
```xml
<insecureSkipHostcheck>true</insecureSkipHostcheck>
```
inside the 
```xml
</gui>
```
node solves this problem.
Filepath to config:
c:\Users\XXX\AppData\Local\Syncthing\config.xml
